### PR TITLE
Fix wrong `DataContext` for multiple mappings

### LINKED ISSUE #1204: Multiple read mappings to same mesh assign wrong mappingContext to dataContext
**Describe your setup**

Current preCICE develop version

**Describe the problem**

When we allowed multiple read mappings to the same mesh in #849, we forgot to update the following assignment (which still assumes that there is only one read mapping to a certain mesh). 

https://github.com/precice/precice/blob/ebce55ab5c563295909b0179f6610b6e4d4fa529/src/precice/config/ParticipantConfiguration.cpp#L517-L534

A user ran into this problem: 

https://precice.discourse.group/t/participant-reads-zeros-for-mapped-data/994/9

(simplified) preCICE config:

```xml
    <mesh name="Fluid1-Centroids">
      <use-data name="Density-f2" />
      <use-data name="Velocity-f2" />
      <use-data name="Pressure-f2" />
    </mesh>

    <mesh name="Fluid2-Centroids">
      <use-data name="Density-f2" />
      <use-data name="Velocity-f2" />
    </mesh>

    <mesh name="Fluid2-Nodes">
      <use-data name="Pressure-f2" />
    </mesh>

    <participant name="Fluid1">
      <use-mesh name="Fluid1-Centroids" provide="yes" />
      <use-mesh name="Fluid2-Centroids" from="Fluid2" />
      <use-mesh name="Fluid2-Nodes" from="Fluid2" />

      <read-data name="Density-f2" mesh="Fluid1-Centroids" />
      <read-data name="Velocity-f2" mesh="Fluid1-Centroids" />
      <read-data name="Pressure-f2" mesh="Fluid1-Centroids" /> 

      <mapping:nearest-neighbor
        direction="read"
        from="Fluid2-Centroids"
        to="Fluid1-Centroids"
        constraint="consistent" />
      <mapping:nearest-projection
        direction="read"
        from="Fluid2-Nodes"
        to="Fluid1-Centroids"
        constraint="consistent" />
    </participant>
[...]
```

Error message

```
(0) 10:09:48 [config::ParticipantConfiguration]:529 in finishParticipantConfiguration: ERROR: Mesh “Fluid2-Nodes” needs to use data “Density-f2” to allow a read mapping to it. Please add a use-data node with name=“Density-f2” to this mesh.
```

Obviously, this error message is wrong, but more concerning is that also the assignment

https://github.com/precice/precice/blob/ebce55ab5c563295909b0179f6610b6e4d4fa529/src/precice/config/ParticipantConfiguration.cpp#L531

is wrong.

**Step To Reproduce**

Let's try to reproduce the problem with an integration test. A copy of this one should be a good starting point:

https://github.com/precice/precice/blob/ebce55ab5c563295909b0179f6610b6e4d4fa529/src/precice/tests/SerialTests.cpp#L2926

**Additional context**

Of course, we have the same problem for multiple **write** mappings **from** the same mesh.

### PR DESCRIPTION
**Pull request by DavidSCN**
## Main changes of this PR
Resolves #1204 and related to #849.

## Motivation and additional information
In case multiple read mappings target the same mesh 'multiple read to mappings' or multiple write mappings originate from the same mesh 'multiple write from mappings', we raised an erroneous error message and assigned the wrong data context. The logic was now reworked. In the configuration, we iterate now first over all mappings and look for corresponding read data and write data tags. The error is now transformed into an if condition and an error is raised in case no proper tags can be found.

## Author's checklist

* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: tests/serial/multiple-mappings/MultipleWriteFromMappings.xml
Comment (DavidSCN) [inline tests/serial/multiple-mappings/MultipleWriteFromMappings.xml:40]: @uekerman I'm not sure if this test is supposed to work as well. Currently (as before) it fails. I guess the main reason is that we call 
https://github.com/precice/precice/blob/7618982a33a340a151c6964430799048314d3bdd/src/precice/config/ParticipantConfiguration.cpp#L535
twice using the same data context (although we do the same for the read case, not completely sure about this).

Comment (DavidSCN) [reply tests/serial/multiple-mappings/MultipleWriteFromMappings.xml:40]: > I guess the main reason is that we call

Yes, that's definitely the case.

> although we do the same for the read case where is works, not completely sure about this

No, in this case we configure another data context (namely the read data context, which is different for each mesh and data). The config here simply doesn't work with the `dataContext`

Comment (uekerman) [reply tests/serial/multiple-mappings/MultipleWriteFromMappings.xml:40]: Such a case should work, yes. Good example. We need to extend the implementation here.

**Review by uekerman — CHANGES_REQUESTED** [review]
Good extension of test cases and I like the comments in the configuration part :+1:
The tricky part, however, still remains :grin:

File: src/precice/config/ParticipantConfiguration.cpp
Comment (DavidSCN) [reply src/precice/config/ParticipantConfiguration.cpp:546]: I don't think this is the case. We first iterate over the mappings and look for corresponding data contexts. The data contexts are created according to the read data and write data tag, and not according to the use data tag

File: src/precice/tests/DataContextTest.cpp
Comment (BenjaminRodenberg) [inline src/precice/tests/DataContextTest.cpp:78]: Can you put this into an individual test? This keeps the existing test simple. Each unit test should also have a single purpose.

File: src/precice/tests/DataContextTest.cpp
Comment (BenjaminRodenberg) [inline src/precice/tests/DataContextTest.cpp:130]: Do we also need a test for multiple read mappings?

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:82]: Please also update the class documentation above that a data context can have no, one or multiple mappings.

File: src/precice/impl/DataContext.cpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.cpp:102]: `std::for_each` is nice, but since you already have a loop below, I would directly do the call there.

Comment (BenjaminRodenberg) [reply src/precice/impl/DataContext.cpp:102]: Alternatively, you could also remove the loop and put everything into your lambda function like above with the `std::any_of` in `isMappingRequired`.

File: src/precice/impl/DataContext.cpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.cpp:68]: This avoids by construction that a mapping in the `DataContext` is just overwritten with another mapping accidentally. I'm missing this security mechanism a bit in the new implementation. Is it, for example, allowed that the same mapping is added twice? What happens, if you do this? A corresponding unit test might also be a good idea here.

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:94]: Please update the documentation and name. It should read `_mappingContexts`.

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:106]: This comment is misleading. You don't add a `_mappingContext`, but you add a `mappingContext` to the `_mappingContexts` container, if I understand this correctly.

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:108]: This also needs an update.

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:112]: I slightly prefer `appendMapping`, because this tells me that there is something listlike which contains mappings and I'm appending my mapping to this thing.

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:135]: Please update this also to reflect that we get the ID from a vector.

File: src/precice/impl/DataContext.hpp
Comment (BenjaminRodenberg) [inline src/precice/impl/DataContext.hpp:144]: See above.

File: src/precice/impl/DataContext.cpp
Comment (DavidSCN) [reply src/precice/impl/DataContext.cpp:68]: You cannot override things in this function as `emplace_back` never overrides any content. If you add the same mapping twice, you would also execute it twice. We could safeguard against this if that's what you have in mind.

File: src/precice/tests/DataContextTest.cpp
Comment (DavidSCN) [reply src/precice/tests/DataContextTest.cpp:130]: No, this would be ambiguous from the user perspective and it is also ensured via an assertion.

**Comment by DavidSCN** [comment]
> I don't really like it that the complexity of the configuration is increased, but I think this is just necessary. Or do you by chance have an idea how this could be shifted?

You mean the configuration in ParticipantConfiguration? This was the actual bug which lead downstream to further code changes. The increase in complexity is to some extend the price for the capability of preCICE to handle such cases. For now, I would go with the current version until we have tests covering everything properly. Maybe one can find simplifications later on.

File: docs/changelog/1267.md
Comment (fsimonis) [inline docs/changelog/1267.md:1]: No PR number needed.

```suggestion
- Fixed the wrong data assignment for multiple read mappings to the same mesh or multiple write mappings from the same mesh.
```

File: src/precice/impl/DataContext.cpp
Comment (BenjaminRodenberg) [reply src/precice/impl/DataContext.cpp:68]: I guess it does not make a lot of sense to add the same mapping twice, does it? If you can add a safeguard here, this would be good.

File: src/precice/impl/DataContext.cpp
Comment (DavidSCN) [reply src/precice/impl/DataContext.cpp:102]: I went for the first, as I need the index in the loop.

**Comment by DavidSCN** [comment]
@BenjaminRodenberg I tried to fix all the documentation related things. Please comment if there are still unclear.

File: src/precice/config/ParticipantConfiguration.cpp
Comment (uekerman) [reply src/precice/config/ParticipantConfiguration.cpp:546]:
```suggestion
                  "Participant \"{}\" defines a write mapping from mesh \"{}\" to mesh \"{}\", "
                  "but there is either no corresponding write-data tag or the meshes used " 
                  "by this participant lack the necessary use-data tags.",
                  participant->getName(), mappingContext.mapping->getInputMesh()->getName(), mappingContext.mapping->getOutputMesh()->getName());
```

**Review by BenjaminRodenberg — APPROVED** [review]
Looks good. I found a few more things, but they are not critical and from my Waveform/DataContext perspective I think you can just merge this PR, when you have taken a look at my last remarks.

File: src/precice/impl/DataContext.hpp
Comment (DavidSCN) [reply src/precice/impl/DataContext.hpp]: I don't have a strong opinion on this, so I just changed to the append version.

File: src/precice/impl/DataContext.cpp
Comment (DavidSCN) [reply src/precice/impl/DataContext.cpp:103]: I really like the concise coding style of the version above. It helps me to understand faster what's going on. Now that you leave it to me, I will stick to it.

Comment (BenjaminRodenberg) [reply src/precice/impl/DataContext.cpp:103]: Sounds fair ;)

**Pull request merged**
